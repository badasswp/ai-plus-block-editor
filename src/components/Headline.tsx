import { __ } from '@wordpress/i18n';
import { check } from '@wordpress/icons';
import { select, dispatch } from '@wordpress/data';
import { useState, useEffect } from '@wordpress/element';
import { Button, Icon, TextareaControl } from '@wordpress/components';
import apiFetch from '@wordpress/api-fetch';

import Toast from '../components/Toast';

/**
 * Headline.
 *
 * This Component returns the Headline
 * label and button.
 *
 * @since 1.1.0
 *
 * @returns {JSX.Element}
 */
const Headline = (): JSX.Element => {
  const [ headline, setHeadline ] = useState( '' );
  const [ isLoading, setIsLoading ] = useState( false );
  const { editPost } = dispatch( 'core/editor' ) as any;
  const {
    getCurrentPostId,
    getEditedPostAttribute,
    getEditedPostContent,
  } = select( 'core/editor' );

  const content = getEditedPostContent();

  useEffect( () => {
    setHeadline( getEditedPostAttribute( 'meta' )['apbe_headline'] );
  }, [] )

  /**
   * This function fires when the user clicks
   * the `Generate` button.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleClick = async (): Promise<void> => {
    setIsLoading( true );

    const response = await apiFetch(
      {
        path: '/ai-plus-block-editor/v1/sidebar',
        method: 'POST',
        data: {
          id: getCurrentPostId(),
          text: content.text || content,
          feature: 'headline'
        },
      }
    );

    const { data } = response as any;
    const caption = data.trim().replace( /^"|"$/g, '' );

    let limit = 1;

    const showAnimatedAiText = setInterval( () => {
      if ( limit === caption.length ) {
        clearInterval( showAnimatedAiText );
      }
      setHeadline( caption.substring( 0, limit ) );
      limit++;
    }, 5 )

    setIsLoading( false );
  }

  /**
   * This function fires when the user selects
   * the AI generated result.
   *
   * @since 1.1.0
   *
   * @returns { void }
   */
  const handleSelection = (): void => {
    let limit = 1;

    const showAnimatedAiText = setInterval( () => {
      if ( limit === headline.length ) {
        clearInterval( showAnimatedAiText );
      }
      const editorHeadline = document.querySelector( '.editor-post-title' );
      editorHeadline.innerHTML = headline.substring( 0, limit );
      limit++;
    }, 5 )

    editPost( { title: headline } );
    editPost( { meta: { apbe_headline: headline } } );
  }

  return (
    <>
      <p><strong>{ __( 'Headline', 'ai-plus-block-editor' ) }</strong></p>
      <TextareaControl
        rows={ 4 }
        value={ headline }
        onChange={ e => console.log(e) }
      />
      <div className="apbe-button-group">
        <Button
          variant="primary"
          onClick={ handleClick }
        >
          { __( 'Generate', 'ai-plus-block-editor' ) }
        </Button>
        <Button
          variant="secondary"
          onClick={ handleSelection }
        >
          <Icon icon={ check } />
        </Button>
      </div>
      <Toast
        message={ __( 'AI is generating text, please hold on for a bit...' ) }
        isLoading={ isLoading }
      />
    </>
  )
}

export default Headline;
