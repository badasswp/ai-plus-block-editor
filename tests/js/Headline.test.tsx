import '@testing-library/jest-dom';
import { act, render, fireEvent, waitFor } from '@testing-library/react';

import apiFetch from '@wordpress/api-fetch';
import { useSelect, useDispatch } from '@wordpress/data';

import Headline from '../../src/components/Headline';

jest.mock( '@wordpress/i18n', () => ( {
	__: jest.fn( ( arg ) => arg ),
} ) );

jest.mock( '@wordpress/data', () => ( {
	useSelect: jest.fn(),
	useDispatch: jest.fn(),
} ) );

jest.mock( '@wordpress/notices', () => ( {
	store: 'core/notices',
} ) );

jest.mock( '@wordpress/components', () => ( {
	Button: jest.fn( ( { children, variant, onClick } ) => {
		return (
			<>
				<button className={ variant } onClick={ onClick }>
					{ children }
				</button>
			</>
		);
	} ),

	Icon: jest.fn( () => {
		return <>Icon</>;
	} ),

	TextareaControl: jest.fn( ( { rows, value, onChange } ) => {
		return (
			<>
				<textarea
					rows={ rows }
					onChange={ ( e ) => onChange( e.target.value ) }
					value={ value }
				/>
			</>
		);
	} ),
} ) );

jest.mock( '@wordpress/api-fetch', () => jest.fn() );

describe( 'Headline', () => {
	beforeEach( () => {
		( useSelect as jest.Mock ).mockReturnValue( {
			postId: 1,
			postContent: 'Hello World',
			postHeadline: 'AI generated headline...',
			notices: [],
		} );

		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: jest.fn(),
			savePost: jest.fn(),
			removeNotice: jest.fn(),
			createErrorNotice: jest.fn(),
		} );

		jest.spyOn( console, 'error' ).mockImplementation( () => {} );
	} );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'renders Headline component', () => {
		const { container, getByText, getByRole } = render( <Headline /> );

		expect( container ).toMatchSnapshot();

		expect( getByText( 'AI generated headline...' ) ).toBeInTheDocument();
		expect( getByText( 'Headline' ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Icon' } ) ).toHaveClass(
			'secondary'
		);
		expect( getByRole( 'button', { name: 'Generate' } ) ).toBeVisible();
		expect( getByRole( 'button', { name: 'Generate' } ) ).toHaveClass(
			'primary'
		);
	} );

	it( 'renders fetched API data from AI LLM', async () => {
		const mockEditPost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () => Promise.resolve( 'What a Wonderful World!' ) )
		);

		const { getByText, getByRole } = render( <Headline /> );

		expect( getByText( 'AI generated headline...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'What a Wonderful World!' )
			).toBeInTheDocument();
		} );
	} );

	it( 'renders error notice on API fail', async () => {
		const mockCreateErrorNotice = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			createErrorNotice: mockCreateErrorNotice,
		} );

		( apiFetch as unknown as jest.Mock ).mockRejectedValueOnce(
			new Error( 'AI LLM down...' )
		);

		const { getByText, getByRole } = render( <Headline /> );

		expect( getByText( 'AI generated headline...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockCreateErrorNotice ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'AI generated headline...' )
			).toBeInTheDocument();
		} );
	} );

	it( 'saves the selected AI Headline', async () => {
		const mockEditPost = jest.fn();
		const mockSavePost = jest.fn();
		( useDispatch as jest.Mock ).mockReturnValue( {
			editPost: mockEditPost,
			savePost: mockSavePost,
		} );

		( apiFetch as unknown as jest.Mock ).mockImplementation(
			jest.fn( () => Promise.resolve( 'What a Wonderful World!' ) )
		);

		const { getByText, getByRole } = render( <Headline /> );

		expect( getByText( 'AI generated headline...' ) ).toBeInTheDocument();

		const button = getByRole( 'button', { name: 'Generate' } );
		await act( async () => {
			fireEvent.click( button );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 1 );
			expect(
				getByText( 'What a Wonderful World!' )
			).toBeInTheDocument();
		} );

		const icon = getByRole( 'button', { name: 'Icon' } );
		await act( async () => {
			fireEvent.click( icon );
		} );

		await waitFor( () => {
			expect( mockEditPost ).toHaveBeenCalledTimes( 25 );
			expect( mockSavePost ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
