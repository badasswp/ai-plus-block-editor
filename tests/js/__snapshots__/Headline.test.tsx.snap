// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headline renders Headline component 1`] = `
<div>
  <p>
    <strong>
      Headline
    </strong>
  </p>
  <textarea
    rows="4"
  >
    AI generated headline...
  </textarea>
  <div
    class="apbe-button-group"
  >
    <button
      class="primary"
    >
      Generate
    </button>
    <button
      class="secondary"
    >
      Icon
    </button>
  </div>
</div>
`;
